server/keyspace_service.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2022 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"path"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pingcap/kvproto/pkg/keyspacepb"
+	"github.com/pingcap/kvproto/pkg/pdpb"
+	"github.com/tikv/pd/server/keyspace"
+	"github.com/tikv/pd/server/storage/endpoint"
+	"go.etcd.io/etcd/clientv3"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+// KeyspaceServer wraps GrpcServer to provide keyspace service.
+type KeyspaceServer struct {
+	*GrpcServer
```

Comment:
Why not
```suggestion
	*Server
```

Reply:
Wraping the `GrpcServer` instead of `Server` enables `KeyspaceServer` to reuse many `GrpcServer`'s utility functions, such as `wrapErrorToHeader`, otherwise the same logic will have to be repeated.
This also make it possible to use `unaryMiddleware` to redirect request it's deemed necessary.

Also, logically speaking, `KeyspaceServer` is a extension of `GrpcServer`, so I think it's quite natural to embed it.